compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -2008,28 +2008,63 @@ boolean checkTypeAndVarCountConsistency(BLangTupleVariable varNode, BTupleType t
+                    List<BType> tupleTypes = new ArrayList<>();
+                    BArrayType arrayType = (BArrayType) varNode.type;
+                    for (int i = 0; i < arrayType.size; i++) {
+                        tupleTypes.add(arrayType.eType);
+                    }
+                    tupleTypeNode = new BTupleType(tupleTypes);
+                    break;
                 default:
                     dlog.error(varNode.pos, DiagnosticErrorCode.INVALID_TUPLE_BINDING_PATTERN_DECL, varNode.type);
                     return false;
             }
         }
 
-        if (tupleTypeNode.tupleTypes.size() != varNode.memberVariables.size()
-                || (tupleTypeNode.restType == null && varNode.restVariable != null)
-                ||  (tupleTypeNode.restType != null && varNode.restVariable == null)) {
+        if ((varNode.isDeclaredWithVar && ((varNode.restVariable != null &&
+                 varNode.memberVariables.size() > tupleTypeNode.tupleTypes.size()) ||
+                (varNode.restVariable == null && (tupleTypeNode.tupleTypes.size() != varNode.memberVariables.size()
+                        || tupleTypeNode.restType != null)))) ||
+                (!varNode.isDeclaredWithVar && (tupleTypeNode.tupleTypes.size() != varNode.memberVariables.size() ||
+                        tupleTypeNode.restType == null && varNode.restVariable != null ||
+                        tupleTypeNode.restType != null && varNode.restVariable == null))) {
             dlog.error(varNode.pos, DiagnosticErrorCode.INVALID_TUPLE_BINDING_PATTERN);
             return false;
         }
 
         int ignoredCount = 0;
         List<BLangVariable> memberVariables = new ArrayList<>(varNode.memberVariables);
+        int tupleNodeMemCount = tupleTypeNode.tupleTypes.size();
+        int varNodeMemCount = varNode.memberVariables.size();
+        BType restType = tupleTypeNode.restType;
         if (varNode.restVariable != null) {
             memberVariables.add(varNode.restVariable);
+            if (varNodeMemCount < tupleNodeMemCount) {
+                LinkedHashSet<BType> varTypes = new LinkedHashSet<>();
+                for (int i = varNodeMemCount; i < tupleNodeMemCount; i++) {
+                    varTypes.add(tupleTypeNode.tupleTypes.get(i));
+                }
+                if (restType != null) {
+                    varTypes.add(restType);
+                }
+                restType = BUnionType.create(null, varTypes);
+            }
         }
-        for (int i = 0; i < memberVariables.size(); i++) {
+        int memVarCount = memberVariables.size();
+        for (int i = 0; i < memVarCount; i++) {
             BLangVariable var = memberVariables.get(i);
-            BType type = (i <= tupleTypeNode.tupleTypes.size() - 1) ? tupleTypeNode.tupleTypes.get(i) :
-                    new BArrayType(tupleTypeNode.restType);
+            BType type;
+            if (varNode.restVariable != null && i == memVarCount - 1) {
```

Comment:
We don't have to repeat the `varNode.restVariable != null` check for each iteration, right? Shall we extract this out to a variable outside the loop?

Reply:
Wouldn't it be better to not add the restVariable to member variables and handle it separately? We are unnecessarily repeating this check for each member variable, right?